From e640ede8e7fd4b43a92eaa8a681c4e2e825e7a54 Mon Sep 17 00:00:00 2001
From: Hood Chatham <roberthoodchatham@gmail.com>
Date: Tue, 22 Oct 2024 15:16:03 +0200
Subject: [PATCH 3/6] Add fallback trampoline for JSPI with no type reflection

This patch removes the logic in `_PyEM_detect_type_reflection` leaving it to
`stack_switching.ts` to decide what to do. We remove this for two reasons:
1. Filing `PyEm_CountArgs` from a C function that's only called once breaks
   memory snapshots. This we could fix by moving the logic of
   `_PyEM_detect_type_reflection` to top level, and
2. If stack switching is available we need `_PyEM_detect_type_reflection` to
   return `true` even the wasm type reflection is not present. We added a
   fallback that detects in a wasm 1.0 compatible way.
---
 Python/emscripten_trampoline.c | 12 +-----------
 1 file changed, 1 insertion(+), 11 deletions(-)

diff --git a/Python/emscripten_trampoline.c b/Python/emscripten_trampoline.c
index 960c6b4a2ef..87cc5e9fbf8 100644
--- a/Python/emscripten_trampoline.c
+++ b/Python/emscripten_trampoline.c
@@ -10,17 +10,7 @@
  * https://github.com/GoogleChromeLabs/wasm-feature-detect/blob/main/src/detectors/type-reflection/index.js
  */
 EM_JS(int, _PyEM_detect_type_reflection, (), {
-    if (!("Function" in WebAssembly)) {
-        return false;
-    }
-    if (WebAssembly.Function.type) {
-        // Node v20
-        Module.PyEM_CountArgs = (func) => WebAssembly.Function.type(wasmTable.get(func)).parameters.length;
-    } else {
-        // Node >= 22, v8-based browsers
-        Module.PyEM_CountArgs = (func) => wasmTable.get(func).type().parameters.length;
-    }
-    return true;
+    return !!Module.PyEM_CountArgs;
 });
 
 void
-- 
2.34.1

